FWCore/src/components/PileupMerge.cpp

```diff
@@ -0,0 +1,117 @@
+#include "PileupMerge.h"
+
+
+DECLARE_ALGORITHM_FACTORY(PileupMerge)
+
+PileupMerge::PileupMerge(const std::string& aName, ISvcLocator* aSvcLoc):
+  GaudiAlgorithm(aName, aSvcLoc),
+  m_store(),
+  m_reader() {
+  declareProperty("pileupGenParticles", m_genParticlesOut, "Generated particles collection (output)");
+  declareProperty("pileupGenVertices", m_genVerticesOut, "Generated vertices collection (output)");
+  declareProperty("pileupTrackHits", m_trackHitsOut, "Track Hit Collection (output)");
+  declareProperty("pileupPosTrackHits", m_posTrackHits, "Positioned Track Hits Collection (output)");
+  declareProperty("pileupCaloHits", m_caloHitsOut, "Calo Hit Collection (output)");
+  declareProperty("pileupPosCaloHits", m_posCaloHits, "Positioned Calo Hits Collection (output)");
+  declareProperty("PileUpTool", m_pileUpTool);
+}
+
+StatusCode PileupMerge::initialize() {
+  currentEvent = 0;
+  StatusCode sc = GaudiAlgorithm::initialize();
+  m_reader.openFile(m_pileupFilename);
+  m_store.setReader(&m_reader);
+  IRndmGenSvc* randSvc = svc<IRndmGenSvc>("RndmGenSvc", true);
+  sc = m_flatDist.initialize(randSvc, Rndm::Flat(0., 1.));
+  return sc;
+}
+
+
+StatusCode PileupMerge::finalize() {
+  StatusCode sc = GaudiAlgorithm::finalize();
+  return sc;
+}
+
+
+StatusCode PileupMerge::execute() {
+  unsigned nEvents = m_reader.getEntries();
+
+
+  const fcc::TrackHitCollection* trackHitCollection;
+  fcc::TrackHitCollection* hitCollOut = new fcc::TrackHitCollection();
+
+  const fcc::PositionedTrackHitCollection* posTrackHitCollection;
+  fcc::PositionedTrackHitCollection* posHitCollOut = new fcc::PositionedTrackHitCollection();
+
+  const fcc::MCParticleCollection* mcParticleCollection;
+  fcc::MCParticleCollection* mcPartCollOut = new fcc::MCParticleCollection();
+
+  const fcc::GenVertexCollection* genVertexCollection;
+  fcc::GenVertexCollection* genVertCollOut = new fcc::GenVertexCollection();
+
+  const fcc::CaloHitCollection* caloHitCollection;
+  fcc::CaloHitCollection* caloHitCollOut = new fcc::CaloHitCollection();
+
+  const fcc::PositionedCaloHitCollection* posCaloHitCollection;
+  fcc::PositionedCaloHitCollection* posCaloHitCollOut = new fcc::PositionedCaloHitCollection();
+
+  const unsigned int numPileUp = m_pileUpTool->numberOfPileUp();
+  for(unsigned iev = 0; iev < numPileup; ++iev) {
+
+    if ( m_randomizePileup == true ) {
+      if (m_flatDist() < 0.3) { // skip one in three events
+        ++currentEvent;
+      }
+    }
+    currentEvent = (currentEvent + 1) % nEvents; // start over from beginning if necessary
```

Comment:
Oh, Joschka seems to have found the source of the confusion: cloning the particles / hits below must of course happen inside the loop. And `currentEvent` should of course be `m_currentEvent`, to reflect it is a member. A test would have caught this, sorry for not providing one right away! But I guess it is worthwhile to first make the code a little less repetitive by introducing tools.